### DISCUSSION TITLE
State timeline: Null values can't be mapped

### DISCUSSION THREAD

**Discussion by SoMuchForSubtlety** [grafana/grafana#38295] [Dashboards, Panels & Visualizations]
**What happened**:
gaps in prometheus data can't be mapped to a state
![image](https://user-images.githubusercontent.com/15961647/121498231-84076f80-c9dc-11eb-9f5c-028ead26130e.png)
the null+nan mapping does not work
![image](https://user-images.githubusercontent.com/15961647/121498117-6803ce00-c9dc-11eb-8bd1-f895e2356ccd.png)

**What you expected to happen**:
The gaps to show up as "OFFLINE"

**How to reproduce it (as minimally and precisely as possible)**:
Try to create a state timeline for a prometheus metric with gaps

**Environment**:
- Grafana version: 8.0.0
- Data source type & version: prometheus v2.18.1
- OS Grafana is installed on: docker container on ubuntu host
- User OS & Browser: Fedora 34, Firefox 89

**Comment by aocenas** [comment]
Removing the Prometheus label, I assume this is generic config affecting every data source.

**Comment by dprokop** [comment]
Indeed, null value mappings has no effect, confirmed using attached dashboard[1].

I don't think this is a bug tho, a feature request rather. @torkelo @leeoniya is this intentional not to consider null mappings in the state timeline?

If I understand correctly uPlot uses nulls for gaps indication, which means that if we wanted nulls to be mapped we would have to perform null mapping prior to the data hits GraphNG, basically in https://github.com/grafana/grafana/blob/7a9d8e423039e0e34155d2fd653c49900ae47311/public/app/plugins/panel/state-timeline/utils.ts#L258

At the same time, I think this is a nice-to-have feature request, as the nulls are actually correctly rendered indicating the missing state.

However, the question is if we aim to have feature parity with @ryantxu's Discrete panel which, by default, maps null values to N/A https://play.grafana.org/d/P_Ym1TRnz/discrete-panel-nulls?  

What do you think folks?

[1]
```
{
  "annotations": {
    "list": [
      {
        "builtIn": 1,
        "datasource": "-- Grafana --",
        "enable": true,
        "hide": true,
        "iconColor": "rgba(0, 211, 255, 1)",
        "name": "Annotations & Alerts",
        "type": "dashboard"
      }
    ]
  },
  "editable": true,
  "gnetId": null,
  "graphTooltip": 0,
  "id": 479,
  "links": [],
  "panels": [
    {
      "datasource": null,
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "thresholds"
          },
          "custom": {
            "fillOpacity": 70,
            "lineWidth": 0
          },
          "mappings": [
            {
              "options": {
                "0": {
                  "color": "red",
                  "index": 1,
                  "text": "BAD"
                },
                "1": {
                  "color": "purple",
                  "index": 0,
                  "text": "OK"
                },
                "null": {
                  "color": "yellow",
                  "index": 2,
                  "text": "EH"
                }
              },
              "type": "value"
            }
          ],
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "green",
                "value": null
              },
              {
                "color": "red",
                "value": 80
              }
            ]
          }
        },
        "overrides": []
      },
      "gridPos": {
        "h": 9,
        "w": 12,
        "x": 0,
        "y": 0
      },
      "id": 2,
      "options": {
        "alignValue": "left",
        "legend": {
          "displayMode": "list",
          "placement": "bottom"
        },
        "mergeValues": true,
        "rowHeight": 0.9,
        "showValue": "auto",
        "tooltip": {
          "mode": "single"
        }
      },
      "targets": [
        {
          "refId": "A",
          "scenarioId": "csv_metric_values",
          "stringInput": "1,0,null,1,1,1,0,null,null,null,0"
        }
      ],
      "title": "Panel Title",
      "type": "state-timeline"
    }
  ],
  "refresh": "",
  "schemaVersion": 30,
  "style": "dark",
  "tags": [],
  "templating": {
    "list": []
  },
  "time": {
    "from": "now-6h",
    "to": "now"
  },
  "timepicker": {},
  "timezone": "",
  "title": "Timeline nulls",
  "uid": "HnXTToR7z",
  "version": 4
}
```

**Comment by pkolyvas** [comment]
We're in the process of moving Grafana feature requests to [GitHub Discussions](https://community.grafana.com/t/were-moving-grafana-feature-requests-to-github-discussions/50341/3). This issue is being moved to [Discussions](https://github.com/grafana/grafana/discussions) or has been moved from our issues list, depending on where you're seeing this message. 

 We hope to use the discussion thread to continue building use cases, and additional evidence based on engagement/voting, in order to help us prioritize this feature request later on down the road. 

As with all things Grafana, we value your input and participation in this project and appreciate your patience during this transitional period. Looking forward to the discussion!

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). Due to the age and number of responses to this discussion, we are deciding to close it. If this is something you would like to see in Grafana, feel free to open an [issue](https://github.com/grafana/grafana/issues) so the discussion can continue. Thank you!
